Code/Framework/AzCore/AzCore/UnitTest/UnitTest.h

```diff
@@ -226,54 +221,54 @@ namespace UnitTest
             {
                 BusDisconnect();
 
-                if (m_createdAllocator)
-                {
-                    AZ::AllocatorInstance<AZ::OSAllocator>::Destroy(); // used by the bus
-                }
-
-                // At this point, the AllocatorManager should not have any allocators left. If we happen to have any,
-                // we exit the test with an error code (this way the test process does not return 0 and the test run
-                // is considered a failure).
-                AZ::AllocatorManager& allocatorManager = AZ::AllocatorManager::Instance();
-                const int numAllocators = allocatorManager.GetNumAllocators();
-                int invalidAllocatorCount = 0;
-
-                for (int i = 0; i < numAllocators; ++i)
+                // Leak detection. We need to collect the allocators and then shutdown the environment to remove all
+                // variables that are there before we can detect leaks.
+                AZStd::vector<AZ::IAllocator*, AZStd::stateless_allocator> allocators;
                 {
-                    if (!allocatorManager.GetAllocator(i)->IsLazilyCreated())
+                    AZ::AllocatorManager& allocatorManager = AZ::AllocatorManager::Instance();
+                    const int numAllocators = allocatorManager.GetNumAllocators();
+                    allocators.resize(numAllocators);
+                    // Iterate in reverse order since some allocators could depend on others to do garbage collection.
+                    // If allocatorB depends on allocatorA, allocatorA will be registered before into the allocator manager.
+                    for (int i = numAllocators - 1; i >= 0; --i)
                     {
-                        invalidAllocatorCount++;
+                        AZ::IAllocator* allocator = allocatorManager.GetAllocator(i);
+                        allocator->GarbageCollect();
+                        allocators[i] = allocator; // keep the same order so allocators that depend on others are reporter latter
```

Comment:
Why not `reserve` and `push_back` instead of resize and assign.
At the very least the resize doesn't need to construct, so `resize_no_construct` can also be used